backend/src/main/java/fk/prof/backend/model/policy/ZookeeperBasedPolicyStoreAPI.java

```diff
@@ -0,0 +1,232 @@
+        }
+        this.curatorClient = curatorClient;
+        this.policyPath = policyPath;
+        this.initialized = false;
+    }
+
+    private static String decode32(String str) {
+        return new String(BaseEncoding.base32().decode(str), Charset.forName("utf-8"));
+    }
+
+    private static String encode(String str) {
+        return BaseEncoding.base32().encode(str.getBytes(Charset.forName("utf-8")));
+
+    }
+
+    private void populateCacheFromZK() throws Exception {
+        CountDownLatch syncLatch = new CountDownLatch(1);
+        RuntimeException syncException = new RuntimeException();
+
+        //ZK Sync operation always happens async, since this is essential for us to proceed further, using a latch here
+        ZookeeperUtil.sync(curatorClient, policyPath).setHandler(ar -> {
+            if (ar.failed()) {
+                syncException.initCause(ar.cause());
+            }
+            syncLatch.countDown();
+        });
+
+        boolean syncCompleted = syncLatch.await(10, TimeUnit.SECONDS);
+        if (!syncCompleted) {
+            throw new PolicyException("ZK sync operation taking longer than expected", true);
+        }
+        if (syncException.getCause() != null) {
+            throw new PolicyException(syncException, true);
+        }
+        for (String appId : curatorClient.getChildren().forPath(policyPath)) {
+            for (String clusterId : curatorClient.getChildren().forPath(policyPath + DELIMITER + appId)) {
+                for (String procName : curatorClient.getChildren().forPath(policyPath + DELIMITER + appId + DELIMITER + clusterId)) {
+                    String zNodePath = policyPath + DELIMITER + appId + DELIMITER + clusterId + DELIMITER + procName;
+                    PolicyDTO.PolicyDetails policyDetails = PolicyDTO.PolicyDetails.parseFrom(ZookeeperUtil.readLatestSeqZNodeChild(curatorClient, zNodePath));
+                    inMemoryPolicyCache.put(decode32(appId), decode32(clusterId), decode32(procName), policyDetails);
+                }
+            }
+        }
+    }
+
+    void init() throws Exception {
+        synchronized (this) {
+            if (!initialized) {
+                populateCacheFromZK();
+                initialized = true;
+            }
+        }
+    }
+
+    Set<String> getAppIds(String prefix) {
+        if (prefix == null) return new HashSet<>();
+        Set<String> appIds = inMemoryPolicyCache.getAppId();
+        return appIds.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+
+    }
+
+    Set<String> getClusterIds(String appId, String prefix) {
+        if (prefix == null) return new HashSet<>();
+        Set<String> clusterIds = inMemoryPolicyCache.getClusterIds(appId);
+        return clusterIds.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+    }
+
+    Set<String> getProcNames(String appId, String clusterId, String prefix) {
+        if (prefix == null) return new HashSet<>();
```

Comment:
Actually putting checks everywhere is not the way to go. Use can mention it here in method doc, that the params are supposed to be non-null.
Caller will make the checks.